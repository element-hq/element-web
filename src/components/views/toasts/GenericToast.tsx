/*
Copyright 2024 New Vector Ltd.
Copyright 2020, 2021 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import React, { ComponentType, ReactNode } from "react";
import { Button } from "@vector-im/compound-web";

import { XOR } from "../../../@types/common";

export interface IProps {
    description: ReactNode;
    detail?: ReactNode;
    primaryLabel: string;
    PrimaryIcon?: ComponentType<React.SVGAttributes<SVGElement>>;

    onPrimaryClick(): void;
}

interface IPropsExtended extends IProps {
    secondaryLabel: string;
    SecondaryIcon?: ComponentType<React.SVGAttributes<SVGElement>>;
    destructive?: "primary" | "secondary";
    onSecondaryClick(): void;
}

const GenericToast: React.FC<XOR<IPropsExtended, IProps>> = ({
    description,
    detail,
    primaryLabel,
    PrimaryIcon,
    secondaryLabel,
    SecondaryIcon,
    destructive,
    onPrimaryClick,
    onSecondaryClick,
}) => {
    const detailContent = detail ? <div className="mx_Toast_detail">{detail}</div> : null;

    return (
        <div>
            <div className="mx_Toast_description">
                {description}
                {detailContent}
            </div>
            <div className="mx_Toast_buttons" aria-live="off">
                {onSecondaryClick && secondaryLabel && (
                    <Button
                        onClick={onSecondaryClick}
                        kind={destructive === "secondary" ? "destructive" : "secondary"}
                        Icon={SecondaryIcon}
                        size="sm"
                    >
                        {secondaryLabel}
                    </Button>
                )}
                <Button
                    onClick={onPrimaryClick}
                    kind={destructive === "primary" ? "destructive" : "primary"}
                    Icon={PrimaryIcon}
                    size="sm"
                >
                    {primaryLabel}
                </Button>
            </div>
        </div>
    );
};

export default GenericToast;
