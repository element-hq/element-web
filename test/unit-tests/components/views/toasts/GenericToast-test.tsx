/*
Copyright 2024 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import { render, RenderResult } from "jest-matrix-react";
import React, { ComponentProps } from "react";

import GenericToast from "../../../../../src/components/views/toasts/GenericToast";

const renderGenericToast = (props: Partial<ComponentProps<typeof GenericToast>> = {}): RenderResult => {
    const propsWithDefaults = {
        primaryLabel: "Accept",
        description: <div>Description</div>,
        onPrimaryClick: () => {},
        onSecondaryClick: () => {},
        secondaryLabel: "Reject",
        ...props,
    };

    return render(<GenericToast {...propsWithDefaults} />);
};

describe("GenericToast", () => {
    it("should render as expected with detail content", () => {
        const { asFragment } = renderGenericToast();
        expect(asFragment()).toMatchSnapshot();
    });

    it("should render as expected without detail content", () => {
        const { asFragment } = renderGenericToast({
            detail: "Detail",
        });
        expect(asFragment()).toMatchSnapshot();
    });
});
