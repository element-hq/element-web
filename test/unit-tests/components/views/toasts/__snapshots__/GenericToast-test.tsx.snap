// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericToast should render as expected with detail content 1`] = `
<DocumentFragment>
  <div>
    <div
      class="mx_Toast_description"
    >
      <div>
        Description
      </div>
    </div>
    <div
      aria-live="off"
      class="mx_Toast_buttons"
    >
      <button
        class="_button_i91xf_17"
        data-kind="secondary"
        data-size="sm"
        role="button"
        tabindex="0"
      >
        Reject
      </button>
      <button
        class="_button_i91xf_17"
        data-kind="primary"
        data-size="sm"
        role="button"
        tabindex="0"
      >
        Accept
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`GenericToast should render as expected without detail content 1`] = `
<DocumentFragment>
  <div>
    <div
      class="mx_Toast_description"
    >
      <div>
        Description
      </div>
      <div
        class="mx_Toast_detail"
      >
        Detail
      </div>
    </div>
    <div
      aria-live="off"
      class="mx_Toast_buttons"
    >
      <button
        class="_button_i91xf_17"
        data-kind="secondary"
        data-size="sm"
        role="button"
        tabindex="0"
      >
        Reject
      </button>
      <button
        class="_button_i91xf_17"
        data-kind="primary"
        data-size="sm"
        role="button"
        tabindex="0"
      >
        Accept
      </button>
    </div>
  </div>
</DocumentFragment>
`;
